lib/jxl/dec_patch_dictionary.cc

```diff
@@ -177,44 +177,141 @@ int PatchDictionary::GetReferences() const {
-  std::sort(sorted_patches_y.begin(), sorted_patches_y.end());
-  patch_starts_.resize(sorted_patches_y.back().first + 2,
-                       sorted_patches_y.size());
-  sorted_patches_.resize(sorted_patches_y.size());
-  for (size_t i = 0; i < sorted_patches_y.size(); i++) {
-    sorted_patches_[i] = sorted_patches_y[i].second;
-    patch_starts_[sorted_patches_y[i].first] =
-        std::min(patch_starts_[sorted_patches_y[i].first], i);
+  // Create a y-interval for each patch.
+  std::vector<PatchInterval> intervals(positions_.size());
+  for (size_t i = 0; i < positions_.size(); ++i) {
+    const auto& pos = positions_[i];
+    intervals[i].idx = i;
+    intervals[i].y0 = pos.y;
+    intervals[i].y1 = pos.y + ref_positions_[pos.ref_pos_idx].ysize;
   }
-  for (size_t i = patch_starts_.size() - 1; i > 0; i--) {
-    patch_starts_[i - 1] = std::min(patch_starts_[i], patch_starts_[i - 1]);
+  auto sort_by_y0 = [&intervals](size_t start, size_t end) {
+    std::sort(&intervals[start], &intervals[end],
+              [](const PatchInterval& i0, const PatchInterval& i1) {
+                return i0.y0 < i1.y0;
+              });
+  };
+  auto sort_by_y1 = [&intervals](size_t start, size_t end) {
+    std::sort(&intervals[start], &intervals[end],
+              [](const PatchInterval& i0, const PatchInterval& i1) {
+                return i0.y1 < i1.y1;
+              });
+  };
+  // Count the number of patches for each row.
+  sort_by_y1(0, intervals.size());
+  num_patches_.resize(intervals.back().y1);
+  for (auto iv : intervals) {
+    for (size_t y = iv.y0; y < iv.y1; ++y) num_patches_[y]++;
   }
+  PatchTreeNode root;
+  root.start = 0;
+  root.num = intervals.size();
+  patch_tree_.push_back(root);
+  size_t next = 0;
+  while (next < patch_tree_.size()) {
+    auto& node = patch_tree_[next];
+    size_t start = node.start;
+    size_t end = node.start + node.num;
+    // Choose the y_center for this node to be the median of interval starts.
+    sort_by_y0(start, end);
+    size_t middle_idx = start + node.num / 2;
+    node.y_center = intervals[middle_idx].y0;
+    // Divide the intervals in [start, end) into three groups:
+    //   * those completely to the right of y_center: [right_start, end)
+    //   * those overlapping y_center: [left_end, right_start)
+    //   * those completely to the left of y_center: [start, left_end)
+    size_t right_start = middle_idx;
+    while (right_start < end && intervals[right_start].y0 == node.y_center) {
+      ++right_start;
+    }
+    sort_by_y1(start, right_start);
+    size_t left_end = right_start;
+    while (left_end > start && intervals[left_end - 1].y1 > node.y_center) {
+      --left_end;
+    }
+    // Fill in sorted_patches_y0_ and sorted_patches_y1_ for the current node.
+    node.num = right_start - left_end;
+    node.start = sorted_patches_y0_.size();
+    for (int i = (int)right_start - 1; i >= (int)left_end; --i) {
```

Comment:
right_start and `left_end` are `size_t`s, so I'd either use `int64_t`, make them `int`s too if we don't believe more than an int will be useful, or do the loop with `size_t` somehow making sure it works out fine :)

Reply:
Changed to int64_t